src/read_write/raw.rs

```diff
@@ -107,45 +109,228 @@ impl NodeReader for RawNodeReader {
             }
         }
 
-        point.color.red = self.attribute_readers[0].read_u8()?;
-        point.color.green = self.attribute_readers[0].read_u8()?;
-        point.color.blue = self.attribute_readers[0].read_u8()?;
+        if let Some(cr) = self.attribute_readers.get_mut("color") {
+            point.color.red = cr.reader.read_u8()?;
+            point.color.green = cr.reader.read_u8()?;
+            point.color.blue = cr.reader.read_u8()?;
+        }
 
-        if let Some(ir) = self.attribute_readers.get_mut(1) {
-            point.intensity = Some(ir.read_f32::<LittleEndian>()?);
+        if let Some(ir) = self.attribute_readers.get_mut("intensity") {
+            point.intensity = Some(ir.reader.read_f32::<LittleEndian>()?);
         }
 
         Ok(point)
     }
 }
 
+impl RawNodeReader {
+    #[allow(dead_code)]
+    fn read_batch(&mut self) -> Option<PointsBatch> {
```

Comment:
How about generalizing `NodeReader` over the return type `<P>` instead and having two implementations of `read`, one for `Point` and one for `PointsBatch`?
Also, returning a result instead of an option may be more appropriate in case the underlying data is corrupt...

Reply:
What was the reasoning for doing it this way instead of just two methods in the same trait?

Reply:
I agree that the choice of generalization or several functions is a matter of taste, I think the main reason at that time was, that for generalization not all writers had to implement everything, so we could migrate step by step.